subprojects/build-cache/src/main/java/org/gradle/caching/local/internal/H2BuildCacheService.java

```diff
@@ -97,12 +98,14 @@ public boolean load(BuildCacheKey key, BuildCacheEntryReader reader) throws Buil
     public void store(BuildCacheKey key, BuildCacheEntryWriter writer) throws BuildCacheException {
         try (Connection conn = dataSource.getConnection()) {
             try (PreparedStatement stmt = conn.prepareStatement("insert ignore into filestore.catalog(entry_key, entry_size, entry_content) values (?, ?, ?)")) {
-                try (InputStream input = writer.openStream()) {
-                    stmt.setString(1, key.getHashCode());
-                    stmt.setLong(2, writer.getSize());
-                    stmt.setBinaryStream(3, input);
-                    stmt.executeUpdate();
+                stmt.setString(1, key.getHashCode());
+                Blob blob = conn.createBlob();
+                try (OutputStream out = blob.setBinaryStream(1)) {
+                    writer.writeTo(out);
                 }
+                stmt.setLong(2, blob.length());
+                stmt.setBlob(3, blob);
+                stmt.executeUpdate();
```

Comment:
🤔 It seems that this approach works through a rather different code path. When using `JdbcStatement.setBinaryStream()`, we end up here, where it seems the `InputStream` is read fully synchronously:

https://github.com/h2database/h2database/blob/4d8d5c62059939a1395a769da3f9cd56a5e7ddf6/h2/src/main/org/h2/value/ValueBlob.java#L59-L81

When we provide an `OutputStream` via `Blob.setBinaryStream()`, then it sets up a piped input and output stream that will eventually result in the above code being called from a newly created thread:

https://github.com/h2database/h2database/blob/4d8d5c62059939a1395a769da3f9cd56a5e7ddf6/h2/src/main/org/h2/jdbc/JdbcBlob.java#L173-L195

This seems highly inefficient for our usage.